### PR TITLE
[ci] Introduces a model test for Owner

### PR DESCRIPTION
**Pull request by hennevogel**
So far we test this though a controller test. But there are many conditions
for the owner search that are easier tested in a model test.

### CODE REVIEW

**Comment by adrianschroeter** [comment]
I miss the missing owner search here, or maybe I just did not spot it?

You should maybe just cherry-pick my commit for the other pull request and avoid the skip() lines...

**Comment by hennevogel** [comment]
@adrianschroeter what is the "missing owner search"?

File: src/api/spec/models/owner_spec.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      before do
+        allow(Backend::Api::Search).to receive(:binary).and_return(collection)
+      end
+
+      it 'returns results' do
+        subject = Owner.search({}, package).first
+        expect(subject.users).to eq('maintainer'=>['Iggy'])
+      end
+
+      it 'respects User.owner' do
```
Comment (hennevogel) [inline src/api/spec/models/owner_spec.rb]: I guess you mean this?

**Comment by bgeuken** [comment]
@adrianschroeter ping :-)

**Comment by adrianschroeter** [comment]
the test case was wrong, an owner is not modifing the role of the package, it just has an influence on the search result for unmaintained packages. feel free to squash my commit into yours.

**Comment by adrianschroeter** [comment]
still not correct .. give me another try...

**Comment by adrianschroeter** [comment]
I have to admit that this is test is just testing one special situation, with bugowners in packages. However this is the most critical one for us.

**Comment by hennevogel** [comment]
I have a hard time understanding what them spec is testing after 79baf56ed255011dc24e7846e1e7dcbbf5478bd4...

1. `user`s owner is `develuser` but `Owner.search` returns `0` instead of `develuser`
1. Once `develuser` is locked  `Owner.search` returns `1`

Shouldn't this be the other way around?

**Comment by hennevogel** [comment]
Running this manually `Owner.search` returns an `Owner` object without users.

```ruby
Owner
  #search
    owner search in maintenance projects
      respects maintenance project suffixes
    missing owner search in normal projects
/usr/bin/rspec: No such file or directory - which

From: /obs/src/api/spec/models/owner_spec.rb @ line 37 :

    32:         subject = Owner.search({ devel: false, filter: 'bugowner' }, nil)
    33:         expect(subject.count).to eq(0)
    34: 
    35:         # locked devel user makes the package unmaintained
    36:         develuser.update_attributes(state: :locked)
 => 37:         binding.pry
    38:         subject = Owner.search({ devel: false, filter: 'bugowner' }, nil)
    39:         expect(subject.count).to eq(1)
    40:       end
    41:     end
    42: 

[1] pry(#<RSpec::ExampleGroups::Owner::Search::MissingOwnerSearchInNormalProjects>)> subject = Owner.search({ devel: false, filter: 'bugowner' }, nil)
=> [#<Owner:0x0000000008a67c70 @package="TestPack", @project="home:Iggy", @rootproject="home:Iggy">]
[2] pry(#<RSpec::ExampleGroups::Owner::Search::MissingOwnerSearchInNormalProjects>)> subject.first.users
=> nil
```

**Comment by adrianschroeter** [comment]
On Mittwoch, 2. Mai 2018, 13:16:12 CEST wrote Henne Vogelsang:
> I have a hard time understanding what them spec is testing after 79baf56ed255011dc24e7846e1e7dcbbf5478bd4...
> 
> 1. `user`s owner is `develuser` but `Owner.search` returns `0` instead of `develuser`
> 1. Once `develuser` is locked  `Owner.search` returns `1`
> 
> Shouldn't this be the other way around?

The search tries to find packages which lack a proper owner definition.
In this case the only user who has a role "bugowner" got a locked owner, so it does not
qualify anymore.
 
 So we find the package as "missing owner" entry for release and maintenance managers.
They can ensure that package gets a responsible person or group again.

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by hennevogel** [comment]
> The search tries to find packages which lack a proper owner definition.

So depending on the object I pass in this is either a search for users or for packages? Phew...

**Comment by adrianschroeter** [comment]
On Mittwoch, 2. Mai 2018, 16:10:44 CEST wrote Henne Vogelsang:
> > The search tries to find packages which lack a proper owner definition.
> 
> So depending on the object I pass in this is either a search for users or for packages? Phew...

it always returns an Owner object, which is rootproject/project/package + optional users or groups.

yes, the interface is not obvious but it got beautified via refactoring :)

However, this is only an internal interface, it can be changed.


-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by hennevogel** [comment]
> However, this is only an internal interface, it can be changed.

Yes, but first the test. On it..

**Comment by hennevogel** [comment]
@adrianschroeter okay I've adopted the test. Your code is still failing for the [Owner.search](https://github.com/openSUSE/open-build-service/pull/4745/files#diff-3581f966eff6f2f2c7eaf66deb6e7147R27) in objects.

Are you going to fix this?

**Comment by hennevogel** [comment]
@adrianschroeter ping

**Comment by coolo** [comment]
I looked at this situation and I think we don't want owner search to map user owners. These owned accounts are only to have locked accounts not to be displayed as missing. So IMO just delete that part

**Comment by coolo** [comment]
ok, looking into it again with the PO, it's all a big mess. Let me fix this for real :)

File: src/api/spec/models/owner_spec.rb [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+      let!(:kind_patchinfo) { PackageKind.create(package_id: patchinfo_42.id, kind: 'patchinfo') }
+      let(:maintenance_user) { create(:confirmed_user, login: 'MaintenanceIggy') }
+      # FIXME: bugowner should be a transitive argument to the package factory
+      let!(:maintenance_bugowner) { create(:relationship_package_user_as_bugowner, user: maintenance_user, package: patchinfo_42) }
+
+      before do
+        allow(Backend::Api::Search).to receive(:binary).and_return(maintenance_collection)
+      end
+      it 'respects maintenance project suffixes' do
+        expect(Owner.search({}, 'package').first.users).to eq('bugowner'=>['hans'])
```
Comment (coolo) [inline src/api/spec/models/owner_spec.rb]: this is supposed to test search for package not for 'package', right?

Comment (coolo) [reply src/api/spec/models/owner_spec.rb]: scratch that - I didn't realize just how horrible this api is

Comment (hennevogel) [reply src/api/spec/models/owner_spec.rb]: ...

**Pull request closed without merging**